### ISSUE TITLE
how to resume training from a lora checkpoint?

### ISSUE THREAD

**Issue by millix19** [Lightning-AI/litgpt#2021]
as title

**Comment by Borda** [comment]
## 🔄 How to Resume Training from a LoRA Checkpoint in lit-GPT

To resume training from a **LoRA checkpoint** in **lit-GPT**, you need to:

1. **Load the base model**.
2. **Load the LoRA adapters from the checkpoint**.
3. **Reattach the LoRA weights to the model**.
4. **Resume training with the optimizer state (if available)**.

Here’s a step-by-step guide:

---

### ✅ Step 1: Make sure you saved a proper LoRA checkpoint

If you're using `save_lora.py`, you should have a checkpoint directory that includes:

- `lit_model_lora_finetuned.pth`
- `config.json` (or other model config)
- Optionally: optimizer and scheduler state dicts (if you saved them manually)

---

### ✅ Step 2: Load the base model

```python
from lit_gpt import GPT, Config

config = Config.from_name("your-base-model-name")  # e.g. "gpt2"
model = GPT(config)
model.load_state_dict(torch.load("path/to/base/model.pth"))
```

---

### ✅ Step 3: Load and apply LoRA weights

```python
from lit_gpt.lora import mark_only_lora_as_trainable, merge_lora_weights

# Apply LoRA to the model
model = GPT(config)  # reload base
model.load_state_dict(torch.load("path/to/base/model.pth"))

# Load LoRA adapter weights
lora_checkpoint = torch.load("path/to/lit_model_lora_finetuned.pth")
model.load_state_dict(lora_checkpoint, strict=False)

# (Optional) Continue training only LoRA layers
mark_only_lora_as_trainable(model)
```

> ⚠️ If you previously saved merged weights (i.e., `merge_lora_weights(model)` was called), you **cannot resume** training with LoRA layers — they're baked in.

---

### ✅ Step 4: Resume optimizer state (if available)

If you saved optimizer/scheduler states, reload them:

```python
optimizer.load_state_dict(torch.load("path/to/optimizer_state.pth"))
scheduler.load_state_dict(torch.load("path/to/scheduler_state.pth"))
```

---

### ✅ Step 5: Resume training

```python
model.train()
for step in range(start_step, num_training_steps):
    # Training loop logic here
    ...
```

---

### 🔄 Alternative: Use CLI/Script

If you're using `finetune_lora.py`, make sure to:

- Pass the `--resume` flag
- Set `--checkpoint-dir path/to/lora/checkpoint`
- Also pass the base model with `--pretrained`

Example:

```bash
python finetune_lora.py \
  --resume \
  --checkpoint-dir checkpoints/lora-run-001 \
  --pretrained checkpoints/base-model \
  --train-data your_data_path
```
